File: backend/scripts/moderate.mjs
// Credentials come from the environment, never a command-line argument or browser bundle.
const [origin, id, action] = process.argv.slice(2);
if (!origin || !id || !['hide', 'restore'].includes(action) || !process.env.VISITOR_ADMIN_TOKEN) {
  console.error('Usage: npm run slate:moderate -- <city-origin> <slate-id> <hide|restore>\nSet VISITOR_ADMIN_TOKEN in your environment first.');
  process.exit(1);
}
const url = new URL('/api/visitors', origin);
if (url.protocol !== 'https:' && !['localhost', '127.0.0.1'].includes(url.hostname)) throw new Error('Use HTTPS for a hosted city.');
const result = await fetch(url, {
  method: 'PATCH', signal: AbortSignal.timeout(15_000),
  headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${process.env.VISITOR_ADMIN_TOKEN}` },
  body: JSON.stringify({ id, hidden: action === 'hide' }),
});
if (!result.ok) { console.error(`Moderation failed (${result.status}): ${await result.text()}`); process.exit(1); }
console.log(action === 'hide' ? 'Slate hidden. Its position is reserved and can be restored.' : 'Slate restored.');

File: backend/scripts/moderate.py
"""Hide or restore a slate through the city's authenticated API proxy."""
import argparse
import json
import os
from pathlib import Path
from urllib.request import Request, urlopen
from urllib.parse import urlparse
from dotenv import load_dotenv

load_dotenv(Path(__file__).resolve().parents[1] / '.env')
parser = argparse.ArgumentParser()
parser.add_argument('origin')
parser.add_argument('slate_id')
parser.add_argument('action', choices=['hide', 'restore'])
args = parser.parse_args()
token = os.getenv('VISITOR_ADMIN_TOKEN', '')
if len(token) < 32:
    raise SystemExit('Set VISITOR_ADMIN_TOKEN in backend/.env or the environment first.')
url = urlparse(args.origin)
if url.scheme != 'https' and url.hostname not in ('localhost', '127.0.0.1'):
    raise SystemExit('Hosted moderation requires HTTPS.')
request = Request(args.origin.rstrip('/') + '/api/visitors', method='PATCH',
                  headers={'Content-Type': 'application/json', 'Authorization': 'Bearer ' + token},
                  data=json.dumps({'id': args.slate_id, 'hidden': args.action == 'hide'}).encode())
with urlopen(request, timeout=15) as response:
    print('Slate hidden (restorable).' if args.action == 'hide' else 'Slate restored.')

